### PR TITLE
chore(iast): re.match getitem aspect

### PR DESCRIPTION
**Pull request by gnufede**
Code security: Aspect for `re.Match` indexed access (`[]`) of matches.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Review by juanjux — APPROVED** [review]

**Review by avara1986 — COMMENTED** [review]
Memory leak tests are needed:

https://github.com/DataDog/dd-trace-py/blob/main/scripts/iast/mod_leak_functions.py#L68

https://github.com/DataDog/dd-trace-py/blob/main/tests/appsec/iast/fixtures/propagation_path.py#L167

File: tests/appsec/iast/aspects/test_re_aspects.py [hunk 1]
```diff
@@ -491,3 +492,50 @@ def test_re_finditer_aspect_not_tainted():
     assert isinstance(res_iterator, typing.Iterator)
     for i in res_iterator:
         assert not is_pyobject_tainted(i)
+
+
+def test_re_match_getitem_aspect_tainted_string_re_object():
```
Comment (avara1986) [inline tests/appsec/iast/aspects/test_re_aspects.py:497]: I think a test to check bytes too are worthy

File: tests/appsec/iast/aspects/test_re_aspects.py [hunk 2]
```diff
@@ -491,3 +492,50 @@ def test_re_finditer_aspect_not_tainted():
...
+    assert is_pyobject_tainted(isaac)
+    assert is_pyobject_tainted(newton)
+
+
+def test_re_match_getitem_aspect_not_tainted_string_re_object():
+    not_tainted_isaac_newton = "Isaac Newton, physicist"
+
+    re_obj = re.compile(r"(\w+) (\w+)")
+
+    re_match = re_match_aspect(None, 1, re_obj, not_tainted_isaac_newton)
```
Comment (avara1986) [inline tests/appsec/iast/aspects/test_re_aspects.py:528]: It's ok a unit test but we need to test the AST replacement too. We need to add a function and patching it like: https://github.com/DataDog/dd-trace-py/blob/main/tests/appsec/iast/aspects/test_add_aspect_fixtures.py#L12

**Review by avara1986 — COMMENTED** [review]
You need to add the re method in:
https://github.com/DataDog/flask-realworld-example-app/blob/master/conduit/security/views.py

**Comment by avara1986** [comment]
Benchmark test?
https://github.com/DataDog/dd-trace-py/blob/main/benchmarks/bm/iast_fixtures/str_methods.py

File: tests/appsec/iast/aspects/test_re_aspects.py (hunk 1 above)
Comment (juanjux) [reply tests/appsec/iast/aspects/test_re_aspects.py:497]: Done!

File: ddtrace/appsec/_iast/_taint_tracking/Aspects/AspectIndex.cpp [hunk 3]
```diff
@@ -48,22 +69,19 @@ api_index_aspect(PyObject* self, PyObject* const* args, const Py_ssize_t nargs)
...
-    }
+    const auto result_o = PyObject_GetItem(candidate_text, idx);
     TRY_CATCH_ASPECT("index_aspect", return result_o, , {
-        const auto ctx_map = Initializer::get_tainting_map();
-        if (not ctx_map or ctx_map->empty()) {
+        if (has_pyerr()) {
+            return nullptr;
+        }
+
+        if ((!is_text(candidate_text) or !is_some_number(idx)) and !PyReMatch_Check(candidate_text)) {
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_taint_tracking/Aspects/AspectIndex.cpp:78]:
```
or !is_some_number(idx))
```
Would this work with named groups? In that case idx is string, no?

Comment (juanjux) [reply ddtrace/appsec/_iast/_taint_tracking/Aspects/AspectIndex.cpp:78]: Yup since is in the parenthesis with the is_text!

**Pull request merged**